File: src/crypto/Keccak256/Keccak256.swift
import Foundation
@_implementationOnly import CVoltaire

/// Keccak-256 hashing functions
public enum Keccak256 {
    /// Compute Keccak-256 hash of data
    public static func hash(_ data: [UInt8]) -> Hash {
        var cHash = PrimitivesHash()
        let rc = data.withUnsafeBufferPointer { ptr in
            primitives_keccak256(ptr.baseAddress, ptr.count, &cHash)
        }
        assert(rc == PRIMITIVES_SUCCESS, "primitives_keccak256 failed: \(rc)")
        return Hash(cHash: cHash)
    }

    /// Compute Keccak-256 hash of data
    public static func hash(_ data: Data) -> Hash {
        hash([UInt8](data))
    }

    /// Compute Keccak-256 hash of UTF-8 string
    public static func hash(_ string: String) -> Hash {
        hash([UInt8](string.utf8))
    }

    /// Hash a message using EIP-191 personal sign format
    public static func hashEthereumMessage(_ message: [UInt8]) -> Hash {
        var out = PrimitivesHash()
        let rc = message.withUnsafeBufferPointer { ptr in
            primitives_eip191_hash_message(ptr.baseAddress, ptr.count, &out)
        }
        assert(rc == PRIMITIVES_SUCCESS, "primitives_eip191_hash_message failed: \(rc)")
        return Hash(cHash: out)
    }

    /// Hash a message using EIP-191 personal sign format
    public static func hashEthereumMessage(_ message: Data) -> Hash {
        hashEthereumMessage([UInt8](message))
    }

    /// Hash a message using EIP-191 personal sign format
    public static func hashEthereumMessage(_ message: String) -> Hash {
        hashEthereumMessage([UInt8](message.utf8))
    }
}

File: src/crypto/Keccak256/Keccak256.kt
package com.voltaire

/** Keccak-256 hashing */
object Keccak256 {
    /** Compute Keccak-256 hash */
    @JvmStatic
    fun hash(data: ByteArray): Hash {
        val out = PrimitivesHash()
        VoltaireLib.INSTANCE.primitives_keccak256(data, data.size.toLong(), out)
        return Hash(out)
    }

    /** Hash a string (UTF-8 encoded) */
    @JvmStatic
    fun hash(data: String): Hash = hash(data.toByteArray(Charsets.UTF_8))

    /** Hash message with EIP-191 personal sign format */
    @JvmStatic
    fun hashMessage(message: ByteArray): Hash {
        val out = PrimitivesHash()
        VoltaireLib.INSTANCE.primitives_eip191_hash_message(message, message.size.toLong(), out)
        return Hash(out)
    }

    /** Hash message string with EIP-191 personal sign format */
    @JvmStatic
    fun hashMessage(message: String): Hash = hashMessage(message.toByteArray(Charsets.UTF_8))
}
